packages/backend-plugin-manager/src/manager/types.ts

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Logger } from 'winston';
+import { Config } from '@backstage/config';
+import {
+  PluginCacheManager,
+  PluginDatabaseManager,
+  PluginEndpointDiscovery,
+  TokenManager,
+  UrlReader,
+} from '@backstage/backend-common';
+import { Router } from 'express';
+import { PluginTaskScheduler, TaskRunner } from '@backstage/backend-tasks';
+import { IdentityApi } from '@backstage/plugin-auth-node';
+import { PermissionEvaluator } from '@backstage/plugin-permission-common';
+import {
+  EventBroker,
+  HttpPostIngressOptions,
+} from '@backstage/plugin-events-node';
+
+import { BackendFeature } from '@backstage/backend-plugin-api';
+import { PackagePlatform, PackageRole } from '@backstage/cli-node';
+import { CatalogBuilder } from '@backstage/plugin-catalog-backend';
+import { TemplateAction } from '@backstage/plugin-scaffolder-node';
+import { IndexBuilder } from '@backstage/plugin-search-backend-node';
+import { EventsBackend } from '@backstage/plugin-events-backend';
+import { PermissionPolicy } from '@backstage/plugin-permission-node';
+
+/**
+ * @public
+ */
+export type PluginEnvironment = {
+  logger: Logger;
+  cache: PluginCacheManager;
+  database: PluginDatabaseManager;
+  config: Config;
+  reader: UrlReader;
+  discovery: PluginEndpointDiscovery;
+  tokenManager: TokenManager;
+  permissions: PermissionEvaluator;
+  scheduler: PluginTaskScheduler;
+  identity: IdentityApi;
+  eventBroker: EventBroker;
+  pluginProvider: BackendPluginProvider;
+};
+
+/**
+ * @public
+ */
+export interface BackendPluginProvider {
+  backendPlugins(): BackendDynamicPlugin[];
+}
+
+/**
+ * @public
+ */
+export interface BaseDynamicPlugin {
+  name: string;
+  version: string;
+  role: PackageRole;
+  platform: PackagePlatform;
+}
+
+/**
+ * @public
+ */
+export type DynamicPlugin = FrontendDynamicPlugin | BackendDynamicPlugin;
```

Comment:
I answered this expected remark in the RFC already (in the *Proposed API* section) with the following note:
> Note:
> Though this RFC is focused towards implementing the support for dynamic backend plugins, front-end plugins have been considered as a discoverable entity api-wise because frontend and backend plugins generally go hand-in-hand to comprise a single feature. We can then envision that, in future steps of the dynamic plugins work, whatever be the way dynamic frontend plugins are implemented and wired into the front-end application, it would make sense for the plugin manager running on the backend to be able to discover them.

Having the backend plugin manager both:

- locate, load and start dynamic backend plugins,
- but also locate dynamic frontend plugins, in order to be able (obviously in the next step) to serve the related assets,

makes sense to me.

cc @tumido

Reply:
I mean sure we can ship it, it doesn't behave badly at this point 😅